Add a test site

danielbrzn opened the pull request:
This PR seeks to add a basic test site for Markbind.

It mainly tests the includes functionality of Markbind such as:
- Nested includes
- Boilerplate includes
- Dynamic includes
- Normal includes
- Segment includes
- HTML includes

In addition, it merges the `example` folder into a unified `test_site` folder that can be used for testing.

Commit:
Add basic test site

danielbrzn committed:
Add include from other Markbind site

- Remove example folder
- Add tooltip text

nicholaschuayunzhi commented:
I have also managed to find a way to do automated testing with Travis when submitting PRs.

https://github.com/nicholaschuayunzhi/markbind-cli/pull/1

It tests for for linting error, and then a file compare between an expected site and a generated site during the test.

EDIT: Should make a new pr or commit on top of this?

### Considerations
- Need to build expected site on unix system due to path separators
- Need to set up travis on this repo
- Currently there is no automated test for windows users locally

Gisonrg commented:
~~Can you send a separate PR to delete the outdated `example` folder? Thanks :D~~

Ignore this, didn't read the description carefully

Gisonrg commented:
@nicholaschuayunzhi 
I think the test site Daniel created could be used in the integration test, a.k.a end-to-end test including system IO to simulate a real site building process. 
While your work (I have seen all the mock objects, good work 👍) could be used as unit test, which provides a fast way to ensure things are work / validate functions are called, etc.
Both of your works would be very helpful in testing 💯

rachx commented:
Can/Should we include known bugs in the test site (e.g. multiple inclusion of a modal)?
It might be easier for testing and debugging (we can all refer to the same copy without creating our own test case) and it can be kept in the future for regression testing.

acjh commented:
@rachx Yes, and:

> - We should differentiate between "issues" (reproducing bugs) and "PRs" (regression testing).

— https://github.com/MarkBind/markbind/issues/121#issuecomment-359710181

danielbrzn committed:
Update to include open bugs section

- Make headings for test cases more prominent
- Define baseUrl for site and subsite

rachx commented:
It will be good to add examples of the supported markdown syntax (radio buttons, code blocks etc) and components (question, popover) etc but that can be done in a separate issue.

Hopefully we can merge this soon as it makes testing more convenient

nicholaschuayunzhi commented on test_site/sub_site/site.json:
```diff
@@ -0,0 +1,16 @@
+  ],
+  "ignore": [
+    "_site/*",
+    "site.json"
```
could we ignore `*md` files as well? I believe our test does not need to check for copied md files

danielbrzn replied:
Sure, will add the required changes.

nicholaschuayunzhi commented:
I have updated the script to diff html differently with a separate script, ignoring path separators.
https://github.com/nicholaschuayunzhi/markbind-cli/pull/2

danielbrzn committed:
Add rule to ignore md files for site build

acjh commented:
- Let's move this into **test/test_site**, so that the unit tests can also be in the same folder in future.
  @nicholaschuayunzhi This will affect the paths in your PRs for *automated testing* and *unit tests*.

- Let's add this to **.gitignore**:
  ```
  # Generated site (MarkBind)
  test/test_site/_site
  ```

acjh commented:
@danielbrzn In case you missed my comment above.

danielbrzn commented:
Sorry about that, will get on it.

danielbrzn committed:
Move test_site to a test folder to allow consolidation of future tests

Add generated site files to .gitignore

Gisonrg reviewed (approved):
Good work!
